File: client/src/Frame.cpp
#include "Frame.h"
#include<iostream>

Frame::Frame(string _stompCommand): stompCommand(_stompCommand+"\n"), headers(""), body(""){}
Frame::Frame() : stompCommand(""), headers(""), body(""){}
void Frame::setStompCommand(string stompCommand){
    this->stompCommand = stompCommand + "\n";

}
void Frame::setHeaders(string headers){
    this->headers = headers+"\n\n";
}
void  Frame::setBody(string body){
    this->body = body + "\n";
}
Frame Frame::sendFrame(string destination, string message, string optionalRecieptId){
    Frame frame = Frame("SEND");
    string head = ("destination:"+destination);
    if(optionalRecieptId == "")
        frame.setHeaders(head);
    else
        frame.setHeaders(head+"\nreceipt-id:"+optionalRecieptId);
    frame.setBody(message);
    return frame;
}
Frame Frame::connectFrame(string login, string passcode, string optionalRecieptId){
    Frame frame = Frame("CONNECT");
    string head = ("accept-version:1.2\nhost:stomp.cs.bgu.ac.il\nlogin:"+login+"\npasscode:"+passcode);
    if(optionalRecieptId == "")
        frame.setHeaders(head);
    else
        frame.setHeaders(head+"\nreceipt-id:"+optionalRecieptId);
    return frame;
}
Frame Frame::subscribeFrame(string destination, int subId, string optionalRecieptId){
    Frame frame = Frame("SUBSCRIBE");
    string head = ("destination:"+destination+"\nid:"+std::to_string(subId));
    if(optionalRecieptId == "")
        frame.setHeaders(head);
    else
        frame.setHeaders(head+"\nreceipt-id:"+optionalRecieptId);
    return frame;
}
Frame Frame::unsubscribeFrame(int subId, string optionalRecieptId){
    Frame frame  = Frame("UNSUBSCRIBE");
    string head = "id:";
    head.append(std::to_string(subId));
    if(optionalRecieptId == "")
        frame.setHeaders(head);
    else{
        frame.setHeaders(head+"\nreceipt-id:"+optionalRecieptId);
    }
    return frame;
}
Frame Frame::disconnectFrame(string recieptId){
    Frame frame = Frame("DISCONNECT");
    frame.setHeaders("reciept:"+recieptId);
    return frame;
}
string Frame::toString(){ 
    //return string(); }
    string output = "";
    output.append(stompCommand + headers + body + "^@" + '\u0000');
    return output;
}

File: server/src/main/java/bgu/spl/net/impl/stomp/Frame.java
package bgu.spl.net.impl.stomp;

public class Frame {

    private String stompCommand;
    private String headers;
    private String body;

    
    public Frame(String stompCommand){
        this.stompCommand = stompCommand+"\n";
    }

    public Frame(){
        this.stompCommand = "";
    }

    public void setStompCommand(String stompCommand){
        this.stompCommand = stompCommand+"\n";
    }

    public void setHeaders(String headers){
        this.headers = headers+"\n\n";
    }

    public void setBody(String body){
        if(this.body == null)
            this.body = body+"\n";
        else{
            this.body += body+"\n";
        }
    }

    public static Frame messageFrame(String destination, int subscription, int messageId){
        Frame frame = new Frame("MESSAGE");
        frame.setHeaders("destination:"+destination+"\nsubscription:"+subscription+"\nmessage-id:"+messageId);
        return frame;
    }

    public static Frame connectedFrame(){
        Frame frame = new Frame("CONNECTED");
        frame.setHeaders("version:1.2");
        return frame;
    }

    public static Frame errorFrame(String frameCaused, String message, String theMessege, String detailedInfo){
        Frame frame = new Frame("ERROR");
        if(frameCaused == "")
            frame.setHeaders("message:"+message+2*'\n'+"The messege:\n----\n"+theMessege+"\n----\n"+detailedInfo);
        else
            frame.setHeaders(frameCaused+"\nmessage:"+message+2*'\n'+"The messege:\n----\n"+theMessege+"\n----\n"+detailedInfo);
        return frame;
    }

    public static Frame receiptFrame(String id){
        Frame frame = new Frame("RECEIPT");
        frame.setHeaders("receipt-id:"+id);
        return frame;
    }

    public static Frame notConnectedError(){
        Frame frame = new Frame("ERROR");
        frame.setHeaders("messege: user not connected");
        return frame;
    }

    public String toString(){
        if(body != null)
            return stompCommand+headers+body+"^@\u0000";
        else
        return stompCommand+headers+"^@\u0000";
    }
}
